### DISCUSSION TITLE
How to parse this data to csv

### DISCUSSION THREAD

**Discussion by NSH531** [PowerShell/PowerShell#17897] [Q&A]
```


device_id                  : eb6f33764ea840cdb4ffb4fafc298790
cid                        : c75e5cdf759c449c9790798e8d089a27
agent_load_flags           : 1
agent_local_time           : 2022-06-14T13:41:29.732Z
agent_version              : 6.39.15316.0
bios_manufacturer          : Phoenix Technologies LTD
bios_version               : 6.00
build_number               : 17763
config_id_base             : 65994753
config_id_build            : 15316
config_id_platform         : 3
cpu_signature              : 132689
external_ip                : 82.166.146.228
mac_address                : 00-50-56-87-18-eb
hostname                   : ANALYST-STATION
first_seen                 : 2022-01-06T11:33:17Z
last_seen                  : 2022-07-06T13:03:22Z
local_ip                   : 10.212.134.205
major_version              : 10
minor_version              : 0
os_version                 : Windows Server 2019
os_build                   : 17763
platform_id                : 0
platform_name              : Windows
policies                   : {@{policy_type=prevention; policy_id=3bd5461363ee4d8da3afdd223c5a6293; applied=True;
                             settings_hash=e5f0e710; assigned_date=2022-05-25T13:02:46.177567235Z;
                             applied_date=2022-05-25T13:04:14.166557278Z; rule_groups=System.Object[]}}
reduced_functionality_mode : no
device_policies            : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=; fim=;
                             firewall=}
groups                     : {feb3f132a4874691b19298745f849954}
group_hash                 : 873dd43d446f34e880ad1a3ba64be6fe38bf661726aeeec2176b1ee9ef8908db
product_type               : 3
product_type_desc          : Server
provision_status           : Provisioned
serial_number              : VMware-42 07 b0 35 c1 c7 f4 83-89 b8 fd 4a be 05 75 bf
service_pack_major         : 0
service_pack_minor         : 0
pointer_size               : 8
status                     : normal
system_manufacturer        : VMware, Inc.
system_product_name        : VMware Virtual Platform
tags                       : {FalconGroupingTags/MSSP, FalconGroupingTags/test}
modified_timestamp         : 2022-07-06T13:03:44Z
meta                       : @{version=8714}
kernel_version             : 10.0.17763.2114

device_id                    : e76098504fdf4e57afd8d43cf45d88c5
cid                          : c75e5cdf759c449c9790798e8d089a27
agent_load_flags             : 1
agent_local_time             : 2022-08-11T06:09:29.909Z
agent_version                : 6.42.15610.0
bios_manufacturer            : innotek GmbH
bios_version                 : VirtualBox
build_number                 : 19043
config_id_base               : 65994753
config_id_build              : 15610
config_id_platform           : 3
cpu_signature                : 526060
detection_suppression_status : unsuppressed
external_ip                  : 77.137.66.79
mac_address                  : 08-00-27-de-7b-66
hostname                     : DANIELTESTS
host_hidden_status           : visible
first_seen                   : 2021-10-17T10:05:38Z
last_seen                    : 2022-08-11T13:45:39Z
local_ip                     : 10.0.2.15
major_version                : 10
minor_version                : 0
os_version                   : Windows 10
os_build                     : 19043
platform_id                  : 0
platform_name                : Windows
policies                     : {@{policy_type=prevention; policy_id=afc838773ae3498c998d74917229856c; applied=True;
                               settings_hash=4249a1df; assigned_date=2022-08-07T10:00:26.570861083Z;
                               applied_date=2022-08-07T10:02:25.368603711Z; rule_groups=System.Object[]}}
reduced_functionality_mode   : no
device_policies              : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=; fim=;
                               firewall=}
groups                       : {ada43906c2634f088b6bb8f2e88e3fa8, c43ee613c1d346dc9e3fc992a5d944ca,
                               ee23cfdebac64c39869635a30df8cb68}
group_hash                   : bdb572eb0f11b1cdfeeb4dfba302c641d10bc97fd289892ebf060fd5c9bd8f73
product_type                 : 1
product_type_desc            : Workstation
provision_status             : Provisioned
serial_number                : 0
service_pack_major           : 0
service_pack_minor           : 0
pointer_size                 : 8
status                       : normal
system_manufacturer          : innotek GmbH
system_product_name          : VirtualBox
tags                         : {FalconGroupingTags/Tests}
modified_timestamp           : 2022-08-11T13:46:42Z
meta                         : @{version=1773}
kernel_version               : 10.0.19043.1826

device_id                   : f180b3d5e0044b87a212e27344f4678f
cid                         : c75e5cdf759c449c9790798e8d089a27
agent_load_flags            : 1
agent_local_time            : 2022-08-14T09:15:40.072Z
agent_version               : 6.42.15610.0
bios_manufacturer           : American Megatrends Inc.
bios_version                : 090008
build_number                : 14393
config_id_base              : 65994753
config_id_build             : 15610
config_id_platform          : 3
cpu_signature               : 329303
external_ip                 : 20.232.40.170
mac_address                 : 00-0d-3a-53-3f-6a
instance_id                 : 2a64f22a-3473-4fc4-8360-4ef2d29d60ca
service_provider            : AZURE
service_provider_account_id : c3dddb9a-d89c-4702-8658-9e7d695c71f7
hostname                    : atlas-ad
first_seen                  : 2022-06-12T15:14:56Z
last_seen                   : 2022-08-14T06:17:11Z
local_ip                    : 10.200.11.200
machine_domain              : atlas.lab
major_version               : 10
minor_version               : 0
os_version                  : Windows Server 2016
os_build                    : 14393
ou                          : {Domain Controllers}
platform_id                 : 0
platform_name               : Windows
policies                    : {@{policy_type=prevention; policy_id=b7f7dd2e280b42298fe4964fcaf90af8; applied=True;
                              settings_hash=7ad8902a; assigned_date=2022-06-12T15:16:49.545528276Z;
                              applied_date=2022-06-12T15:18:26.389771917Z; rule_groups=System.Object[]}}
reduced_functionality_mode  : no
device_policies             : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=;
                              identity-protection=; fim=; firewall=}
managed_apps                : @{identity-protection=}
groups                      : {df5fe73eba494c429fdd1a382001513f}
group_hash                  : da7fddcf75b68d1b4732bd819a0354266464f41a9ec97986131e337b3a831c29
product_type                : 2
product_type_desc           : Domain Controller
provision_status            : Provisioned
serial_number               : 0000-0003-0548-3271-0346-5594-28
service_pack_major          : 0
service_pack_minor          : 0
pointer_size                : 8
site_name                   : Default-First-Site-Name
status                      : normal
system_manufacturer         : Microsoft Corporation
system_product_name         : Virtual Machine
tags                        : {}
modified_timestamp          : 2022-08-14T06:22:41Z
meta                        : @{version=367}
zone_group                  : prod-120093_120064_119958_121815
kernel_version              : 10.0.14393.4583

device_id                   : 0ea7d692a6a8437895b16f5d26a66789
cid                         : c75e5cdf759c449c9790798e8d089a27
agent_load_flags            : 1
agent_local_time            : 2022-08-14T09:16:15.955Z
agent_version               : 6.42.15610.0
bios_manufacturer           : American Megatrends Inc.
bios_version                : 090008
build_number                : 19044
config_id_base              : 65994753
config_id_build             : 15610
config_id_platform          : 3
cpu_signature               : 329303
external_ip                 : 20.232.40.170
mac_address                 : 00-0d-3a-53-3b-b1
instance_id                 : 43c8841d-35d4-4d6c-8715-4444d658ff13
service_provider            : AZURE
service_provider_account_id : c3dddb9a-d89c-4702-8658-9e7d695c71f7
hostname                    : WIN10DSKTP
first_seen                  : 2022-08-07T09:20:37Z
last_seen                   : 2022-08-14T06:18:30Z
local_ip                    : 10.200.1.10
machine_domain              : atlas.lab
major_version               : 10
minor_version               : 0
os_version                  : Windows 10
os_build                    : 19044
ou                          : {}
platform_id                 : 0
platform_name               : Windows
policies                    : {@{policy_type=prevention; policy_id=b7f7dd2e280b42298fe4964fcaf90af8; applied=True;
                              settings_hash=7ad8902a; assigned_date=2022-08-07T09:22:23.993394046Z;
                              applied_date=2022-08-07T09:24:10.32747303Z; rule_groups=System.Object[]}}
reduced_functionality_mode  : no
device_policies             : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=; fim=;
                              firewall=}
groups                      : {}
group_hash                  : e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
product_type                : 1
product_type_desc           : Workstation
provision_status            : Provisioned
serial_number               : 0000-0004-8842-9905-6875-3268-28
service_pack_major          : 0
service_pack_minor          : 0
pointer_size                : 8
site_name                   : Default-First-Site-Name
status                      : normal
system_manufacturer         : Microsoft Corporation
system_product_name         : Virtual Machine
tags                        : {}
modified_timestamp          : 2022-08-14T06:19:44Z
meta                        : @{version=148}
zone_group                  : prod-120093_120064_119958_121815
kernel_version              : 10.0.19044.1288

device_id                  : a22a617de1cd437195bf34c99215afd6
cid                        : c75e5cdf759c449c9790798e8d089a27
agent_load_flags           : 1
agent_local_time           : 2022-08-13T00:01:54.444Z
agent_version              : 6.42.15610.0
bios_manufacturer          : Dell Inc.
bios_version               : A24
build_number               : 19044
config_id_base             : 65994753
config_id_build            : 15610
config_id_platform         : 3
cpu_signature              : 198356
external_ip                : 82.166.184.148
mac_address                : 60-57-18-e7-cc-5a
hostname                   : DESKTOP-PH7A683
first_seen                 : 2022-08-08T11:12:47Z
last_seen                  : 2022-08-14T08:42:46Z
local_ip                   : 20.20.20.4
major_version              : 10
minor_version              : 0
os_version                 : Windows 10
os_build                   : 19044
platform_id                : 0
platform_name              : Windows
policies                   : {@{policy_type=prevention; policy_id=654230faa8cc474a95048caf86e4e47c; applied=True;
                             settings_hash=66a2b36c; assigned_date=2022-08-08T11:20:21.845534548Z;
                             applied_date=2022-08-08T11:22:09.342450085Z; rule_groups=System.Object[]}}
reduced_functionality_mode : no
device_policies            : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=; fim=;
                             firewall=}
groups                     : {09a47c36e47441eb9daeee5ad2a35295}
group_hash                 : 87d0f6f051b0b29f28a1cde002f2e81562fca3a3446439c19a1199ffce55f470
product_type               : 1
product_type_desc          : Workstation
provision_status           : Provisioned
serial_number              : D8J5N32
service_pack_major         : 0
service_pack_minor         : 0
pointer_size               : 8
status                     : normal
system_manufacturer        : Dell Inc.
system_product_name        : Latitude E7450
tags                       : {}
modified_timestamp         : 2022-08-14T08:44:12Z
meta                       : @{version=249}
kernel_version             : 10.0.19044.1889

device_id                  : fcfa8cd17b504f20bb7f9db9710d2f00
cid                        : c75e5cdf759c449c9790798e8d089a27
agent_load_flags           : 1
agent_local_time           : 2022-08-09T20:46:42.875Z
agent_version              : 6.42.15610.0
bios_manufacturer          : Phoenix Technologies LTD
bios_version               : 6.00
build_number               : 17763
config_id_base             : 65994753
config_id_build            : 15610
config_id_platform         : 3
cpu_signature              : 132689
external_ip                : 82.166.146.228
mac_address                : 00-50-56-87-18-eb
hostname                   : ANALYST-STATION
first_seen                 : 2022-07-06T13:34:14Z
last_seen                  : 2022-08-14T08:55:19Z
local_ip                   : 10.30.1.2
major_version              : 10
minor_version              : 0
os_version                 : Windows Server 2019
os_build                   : 17763
platform_id                : 0
platform_name              : Windows
policies                   : {@{policy_type=prevention; policy_id=d4fba0b8b03a4117b460a3ed46c5b6c8; applied=True;
                             settings_hash=416606ef; assigned_date=2022-07-06T13:59:55.35281357Z;
                             applied_date=2022-07-06T14:00:56.834091481Z; rule_groups=System.Object[]}}
reduced_functionality_mode : no
device_policies            : @{prevention=; sensor_update=; device_control=; global_config=; remote_response=; fim=;
                             firewall=}
groups                     : {8d9285525ada4736bbd9368a2006f0ec}
group_hash                 : 60997c0870cbeaad3838a08b74d0bace691bd581fd9ece73b5302b562c975e53
product_type               : 3
product_type_desc          : Server
provision_status           : Provisioned
serial_number              : VMware-42 07 b0 35 c1 c7 f4 83-89 b8 fd 4a be 05 75 bf
service_pack_major         : 0
service_pack_minor         : 0
pointer_size               : 8
status                     : normal
system_manufacturer        : VMware, Inc.
system_product_name        : VMware Virtual Platform
tags                       : {FalconGroupingTags/MSSP}
modified_timestamp         : 2022-08-14T08:56:01Z
meta                       : @{version=1885}
kernel_version             : 10.0.17763.2114

**Comment by kasini3000** [comment]
Some fields are missing:
```
detection_suppression_status : unsuppressed
host_hidden_status           : visible
```
It will cause the field names and field values in csv to be unaligned

I think the best way is:
* use Delimiter “:” split
* insert to nosql db,   eg :  litedb or mongodb.

search : powershell litedb to learn

**Comment by NSH531** [comment]
Really?
Need I to insert this data to nosql?


בתאריך יום שני, 15 באוגוסט 2022, מאת kasini3000 ***@***.***>:

> Some fields are missing:
>
> detection_suppression_status : unsuppressed
>
> host_hidden_status           : visible
>
>
> It will cause the field names and field values in csv to be unaligned
>
> I think the best way is:
>
>    - use Delimiter “:” split
>    - insert to nosql db, eg : litedb or mongodb.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/PowerShell/discussions/17897#discussioncomment-3395511>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNBJWLQS3KNCYK4V7DZYLLVZG5J5ANCNFSM56PQNZ7Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
<https://netanel.ml>

**Comment by ili101** [comment]
* If you like community help in Hebrew you can try asking in https://t.me/IsraelPowershellUserGroup

**Comment by NSH531** [comment]
שבוע טוב :-)

אני רוצה לעדכן אותך שלצערי אני לא יכול בשלב זה להמשיך להעסיק את מנהל השיווק
שלי

שמו אור באבד.

הוא עושה עבודה נהדרת!

אני ממליץ עליו בחום וחושב שהוא יכול להיות נכס עבורך.

אור חי בטורונטו קנדה ועובד בשלט רחוק

הוא בוגר תואר מדעי המחשב מאוניברסיטת תל אביב ויש לו ניסיון של 15 שנה בניהול
קמפיינים בישראל ובעולם.

 ניתן לקבוע פגישה בקישור הבא:

https://calendly.com/tigertest-hungary/meet-or-babad
<https://api.mailbrother.com/mail/click/NnN5QXpjcGFwaExhcm9vcEZVQkZmYnZ3QVZaSGphZHowZzVDRlBkREQybm5jaGxLenNHYytSRlBDdmdxNGtPOVpXejZ6T05XSmFVNHoydEgxUGdIMVY5ZC9yNkZwOEQ1eWJRRGl0MTVVWjhhTVh5bHUvbmRrZXlzakE4bzBJMDQ1ODlzWTZOb1NLRFBBRUZqcnF3d01RPT0=>

אני אשמח שתתאמו פגישה.

אני רוצה שתכירו!!!

המייל של אור:

***@***.***



Sent with MailBrother
<https://chrome.google.com/webstore/detail/mailbrother-html-inserter/obngoldljmnnpggbekneikaohbeflbee>

**Comment by rhubarb-geek-nz** [comment]
A simple attempt, the continuation lines are problematic because of spaces.

```
#!/usr/bin/env pwsh
param (
  $InputFile = 'input.txt'
)

$global:List = New-Object 'System.Collections.ArrayList'
$global:Dict = $Null
$global:ParamName = $Null
$global:ParamValue = $Null
$global:Column = 0

function AddToDict()
{
	if ($global:ParamName -ne $Null)
	{
		if ($global:Dict -eq $Null)
		{
			$global:Dict = New-Object 'System.Collections.Generic.Dictionary[string,string]'
		}

		$global:Dict[$global:ParamName] = $global:ParamValue.Trim()

		$global:ParamName = $Null
	}
}

function AddToList {

	AddToDict

	if ($global:Dict -ne $Null)
	{
		$Count = $global:List.Add($global:Dict)

		$global:Dict = $Null
	}
}

Get-Content -Path $InputFile | ForEach-Object {
	if ($_.Length -eq 0)
	{
		AddToList
	}
	else
	{
		if ($_.StartsWith(' '))
		{
			$global:ParamValue = ($global:ParamValue + $_.Substring((1+$global:Column)))
		}
		else
		{
			AddToDict

			$global:Column = $_.IndexOf(':')

			$global:ParamName = $_.Substring(0,$global:Column).Trim()

			$global:ParamValue = $_.Substring((2+$global:Column))
		}		
	}
}

AddToList

$global:List | ConvertTo-Json -Depth 4

```

**Comment by Buzman01** [comment]
What is a CSV? 

CSV (Comma Separated Values) is a simple file format used to store tabular data, such as a spreadsheet or database. A CSV file stores tabular data (numbers and text) in plain text. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. The use of the comma as a field separator is the source of the name for this file format.
For working CSV files in Python, there is an inbuilt module called csv. 

Working with csv files in Python

Example 1: Reading a CSV file

Python

# importing csv module

import csv 

 

# csv file name

filename = "aapl.csv" 

 

# initializing the titles and rows list

fields = [] 

rows = [] 

 

# reading csv file

with open(filename, 'r') as csvfile: 

    # creating a csv reader object 

    csvreader = csv.reader(csvfile) 

     

    # extracting field names through first row 

    fields = next(csvreader) 

 

    # extracting each data row one by one 

    for row in csvreader: 

        rows.append(row) 

 

    # get total number of rows 

    print("Total no. of rows: %d"%(csvreader.line_num)) 

 

# printing the field names

print('Field names are:' + ', '.join(field for field in fields)) 

 

# printing first 5 rows

print('\nFirst 5 rows are:\n') 

for row in rows[:5]: 

    # parsing each column of a row 

    for col in row: 

        print("%10s"%col,end=" "), 

    print('\n')


Working with csv files in Python

This article explains how to load and parse a CSV file in Python.

What is a CSV? 

CSV (Comma Separated Values) is a simple file format used to store tabular data, such as a spreadsheet or database. A CSV file stores tabular data (numbers and text) in plain text. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. The use of the comma as a field separator is the source of the name for this file format.
For working CSV files in Python, there is an inbuilt module called csv. 

Working with csv files in Python

Example 1: Reading a CSV file

Python

# importing csv module

import csv 

 

# csv file name

filename = "aapl.csv" 

 

# initializing the titles and rows list

fields = [] 

rows = [] 

 

# reading csv file

with open(filename, 'r') as csvfile: 

    # creating a csv reader object 

    csvreader = csv.reader(csvfile) 

     

    # extracting field names through first row 

    fields = next(csvreader) 

 

    # extracting each data row one by one 

    for row in csvreader: 

        rows.append(row) 

 

    # get total number of rows 

    print("Total no. of rows: %d"%(csvreader.line_num)) 

 

# printing the field names

print('Field names are:' + ', '.join(field for field in fields)) 

 

# printing first 5 rows

print('\nFirst 5 rows are:\n') 

for row in rows[:5]: 

    # parsing each column of a row 

    for col in row: 

        print("%10s"%col,end=" "), 

    print('\n') 

Output:



￼￼

 

The above example uses a CSV file aapl.csv which can be downloaded from here. 

Run this program with the aapl.csv file in the same directory.

Let us try to understand this piece of code. 

with open(filename, 'r') as csvfile: csvreader = csv.reader(csvfile)

Here, we first open the CSV file in READ mode. The file object is named as csvfile. The file object is converted to csv.reader object. We save the csv.reader object as csvreader.

fields = csvreader.next()

csvreader is an iterable object. Hence, .next() method returns the current row and advances the iterator to the next row. Since the first row of our csv file contains the headers (or field names), we save them in a list called fields.

for row in csvreader: rows.append(row)

Now, we iterate through the remaining rows using a for loop. Each row is appended to a list called rows. If you try to print each row, one can find that a row is nothing but a list containing all the field values.

print("Total no. of rows: %d"%(csvreader.line_num))

csvreader.line_num is nothing but a counter which returns the number of rows that have been iterated.

Example 2: Writing to a CSV file

Python

# importing the csv module

import csv 

 

# field names

fields = ['Name', 'Branch', 'Year', 'CGPA'] 

 

# data rows of csv file

rows = [ ['Nikhil', 'COE', '2', '9.0'], 

        ['Sanchit', 'COE', '2', '9.1'], 

        ['Aditya', 'IT', '2', '9.3'], 

        ['Sagar', 'SE', '1', '9.5'], 

        ['Prateek', 'MCE', '3', '7.8'], 

        ['Sahil', 'EP', '2', '9.1']] 

 

# name of csv file

filename = "university_records.csv" 

 

# writing to csv file

with open(filename, 'w') as csvfile: 

    # creating a csv writer object 

    csvwriter = csv.writer(csvfile) 

     

    # writing the fields 

    csvwriter.writerow(fields) 

     

    # writing the data rows 

    csvwriter.writerows(rows) 

Let us try to understand the above code in pieces. 

fields and rows have been already defined. fields is a list containing all the field names. rows is a list of lists. Each row is a list containing the field values of that row.

with open(filename, 'w') as csvfile: csvwriter = csv.writer(csvfile)

Here, we first open the CSV file in WRITE mode. The file object is named as csvfile. The file object is converted to csv.writer object. We save the csv.writer object as csvwriter.

csvwriter.writerow(fields)

Now we use writerow method to write the first row which is nothing but the field names. 

csvwriter.writerows(rows)

We use writerows method to write multiple rows at once.

Example 3: Writing a dictionary to a CSV file

Python

# importing the csv module

import csv 

 

# my data rows as dictionary objects

mydict =[{'branch': 'COE', 'cgpa': '9.0', 

     

GEEKSFORGEEKS

Working with csv files in Python

This article explains how to load and parse a CSV file in Python.

What is a CSV? 

CSV (Comma Separated Values) is a simple file format used to store tabular data, such as a spreadsheet or database. A CSV file stores tabular data (numbers and text) in plain text. Each line of the file is a data record. Each record consists of one or more fields, separated by commas. The use of the comma as a field separator is the source of the name for this file format.
For working CSV files in Python, there is an inbuilt module called csv. 

Working with csv files in Python

Example 1: Reading a CSV file

Python

# importing csv module

import csv 

 

# csv file name

filename = "aapl.csv" 

 

# initializing the titles and rows list

fields = [] 

rows = [] 

 

# reading csv file

with open(filename, 'r') as csvfile: 

    # creating a csv reader object 

    csvreader = csv.reader(csvfile) 

     

    # extracting field names through first row 

    fields = next(csvreader) 

 

    # extracting each data row one by one 

    for row in csvreader: 

        rows.append(row) 

 

    # get total number of rows 

    print("Total no. of rows: %d"%(csvreader.line_num)) 

 

# printing the field names

print('Field names are:' + ', '.join(field for field in fields)) 

 

# printing first 5 rows

print('\nFirst 5 rows are:\n') 

for row in rows[:5]: 

    # parsing each column of a row 

    for col in row: 

        print("%10s"%col,end=" "), 

    print('\n') 

Output:



￼￼

 

The above example uses a CSV file aapl.csv which can be downloaded from here. 

Run this program with the aapl.csv file in the same directory.

Let us try to understand this piece of code. 

with open(filename, 'r') as csvfile: csvreader = csv.reader(csvfile)

Here, we first open the CSV file in READ mode. The file object is named as csvfile. The file object is converted to csv.reader object. We save the csv.reader object as csvreader.

fields = csvreader.next()

csvreader is an iterable object. Hence, .next() method returns the current row and advances the iterator to the next row. Since the first row of our csv file contains the headers (or field names), we save them in a list called fields.

for row in csvreader: rows.append(row)

Now, we iterate through the remaining rows using a for loop. Each row is appended to a list called rows. If you try to print each row, one can find that a row is nothing but a list containing all the field values.

print("Total no. of rows: %d"%(csvreader.line_num))

csvreader.line_num is nothing but a counter which returns the number of rows that have been iterated.

Example 2: Writing to a CSV file

Python

# importing the csv module

import csv 

 

# field names

fields = ['Name', 'Branch', 'Year', 'CGPA'] 

 

# data rows of csv file

rows = [ ['Nikhil', 'COE', '2', '9.0'], 

        ['Sanchit', 'COE', '2', '9.1'], 

        ['Aditya', 'IT', '2', '9.3'], 

        ['Sagar', 'SE', '1', '9.5'], 

        ['Prateek', 'MCE', '3', '7.8'], 

        ['Sahil', 'EP', '2', '9.1']] 

 

# name of csv file

filename = "university_records.csv" 

 

# writing to csv file

with open(filename, 'w') as csvfile: 

    # creating a csv writer object 

    csvwriter = csv.writer(csvfile) 

     

    # writing the fields 

    csvwriter.writerow(fields) 

     

    # writing the data rows 

    csvwriter.writerows(rows) 

Let us try to understand the above code in pieces. 

fields and rows have been already defined. fields is a list containing all the field names. rows is a list of lists. Each row is a list containing the field values of that row.

with open(filename, 'w') as csvfile: csvwriter = csv.writer(csvfile)

Here, we first open the CSV file in WRITE mode. The file object is named as csvfile. The file object is converted to csv.writer object. We save the csv.writer object as csvwriter.

csvwriter.writerow(fields)

Now we use writerow method to write the first row which is nothing but the field names. 

csvwriter.writerows(rows)

We use writerows method to write multiple rows at once.

Example 3: Writing a dictionary to a CSV file

Python

# importing the csv module

import csv 

 

# my data rows as dictionary objects

mydict =[{'branch': 'COE', 'cgpa': '9.0', 

          'name': 'Nikhil', 'year': '2'}, 

        {'branch': 'COE', 'cgpa': '9.1', 

         'name': 'Sanchit', 'year': '2'}, 

        {'branch': 'IT', 'cgpa': '9.3', 

         'name': 'Aditya', 'year': '2'}, 

        {'branch': 'SE', 'cgpa': '9.5', 

         'name': 'Sagar', 'year': '1'}, 

        {'branch': 'MCE', 'cgpa': '7.8', 

         'name': 'Prateek', 'year': '3'}, 

        {'branch': 'EP', 'cgpa': '9.1', 

         'name': 'Sahil', 'year': '2'}] 

 

# field names

fields = ['name', 'branch', 'year', 'cgpa'] 

 

# name of csv file

filename = "university_records.csv" 

 

# writing to csv file

with open(filename, 'w') as csvfile: 

    # creating a csv dict writer object 

    writer = csv.DictWriter(csvfile, fieldnames = fields) 

     

    # writing headers (field names) 

    writer.writeheader() 

     

    # writing data rows 

    writer.writerows(mydict) 

In this example, we write a dictionary mydict to a CSV file.  

with open(filename, 'w') as csvfile: writer = csv.DictWriter(csvfile, fieldnames = fields)

Here, the file object (csvfile) is converted to a DictWriter object. Here, we specify the fieldnames as an argument. 

writer.writeheader()

writeheader method simply writes the first row of your csv file using the pre-specified fieldnames.

writer.writerows(mydict)

writerows method simply writes all the rows but in each row, it writes only the values(not keys).

So, in the end, our CSV file looks like this: 

￼￼

csv file

Consider that a CSV file looks like this in plain text: 

￼￼

university record

We notice that the delimiter is not a comma but a semi-colon. Also, the rows are separated by two newlines instead of one. In such cases, we can specify the delimiter and line terminator.

Example 4: How Store Emails in CSV files ? 

Python3

# importing the csv module

import csv 

 

# field names

fields = ['Name', 'Email'] 

 

# data rows of csv file

rows = [ ['Nikhil', 'nikhil.gfg@gmail.com'], 

        ['Sanchit', 'sanchit.gfg@gmail.com'], 

        ['Aditya', 'aditya.gfg@gmail.com'], 

        ['Sagar', 'sagar.gfg@gmail.com'], 

        ['Prateek', 'prateek.gfg@gmail.com'], 

        ['Sahil', 'sahil.gfg@gmail.com']] 

 

# name of csv file

filename = "email_records.csv" 

 

# writing to csv file

with open(filename, 'w') as csvfile: 

    # creating a csv writer object 

    csvwriter = csv.writer(csvfile) 

     

    # writing the fields 

    csvwriter.writerow(fields) 

     

    # writing the data rows 

    csvwriter.writerows(rows) 

Output:

￼￼

Emails in csv

Explanation:

Here we are importing the csv module and then simply using the same concept of storing the emails in the form of comma-separated entity also with their names. We’re opening the file open() function and specifying that we need that as a csv file and then writing the each column into the csv file using writer object.
